src/shared/tpm2-util.c

```diff
@@ -4003,6 +4065,644 @@ static int tpm2_deserialize(
+
+        _cleanup_free_ void *buf = NULL;
+        r = kdf_kb_hmac_derive(
+                        "COUNTER",
+                        hash_alg_name,
+                        key,
+                        key_len,
+                        label,
+                        strlen(label),
+                        context,
+                        context_len,
+                        /* seed= */ NULL,
+                        /* seed_len= */ 0,
+                        len,
+                        &buf);
+        if (r < 0)
+                return r;
+
+        /* If the number of bits results in a partial byte, the TPM spec requires we zero the unrequested
+         * bits in the MSB (i.e. at index 0). From the spec Part 1 ("Architecture") section on Key
+         * Derivation Function, specifically KDFa():
+         *
+         * "The implied return from this function is a sequence of octets with a length equal to (bits + 7) /
+         * 8. If bits is not an even multiple of 8, then the returned value occupies the least significant
+         * bits of the returned octet array, and the additional, high-order bits in the 0th octet are
+         * CLEAR. The unused bits of the most significant octet (MSO) are masked off and not shifted." */
+        size_t partial = bits % 8;
+        if (partial > 0)
+                ((uint8_t*) buf)[0] &= 0xff >> (8 - partial);
+
+        *ret_key = TAKE_PTR(buf);
+        *ret_key_len = len;
+
+        return 0;
+}
+
+/* KDFe() as defined by the TPM spec. */
+static int tpm2_kdfe(
+              TPMI_ALG_HASH hash_alg,
+              const void *shared_secret,
+              size_t shared_secret_len,
+              const char *label,
+              const void *context_u,
+              size_t context_u_size,
+              const void *context_v,
+              size_t context_v_size,
+              size_t bits,
+              void **ret_key,
+              size_t *ret_key_len) {
+
+        int r;
+
+        assert(shared_secret);
+        assert(label);
+        assert(context_u);
+        assert(context_v);
+        assert(bits > 0);
+        assert(bits <= SIZE_MAX - 7);
+        assert(ret_key);
+        assert(ret_key_len);
+
+        size_t len = (bits + 7) / 8;
+
+        const char *hash_alg_name = tpm2_hash_alg_to_string(hash_alg);
+        if (!hash_alg_name)
+                return log_debug_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE),
+                                       "Could not get hash name for 0x%" PRIx16 ".", hash_alg);
+
+        _cleanup_free_ uint8_t *info = malloc(strlen(label) + 1 + context_u_size + context_v_size);
+        if (!info)
+                return log_oom_debug();
+
+        size_t info_len = 0;
+        memcpy(info, label, strlen(label) + 1);
+        info_len += strlen(label) + 1;
+        memcpy(&info[info_len], context_u, context_u_size);
+        info_len += context_u_size;
+        memcpy(&info[info_len], context_v, context_v_size);
+        info_len += context_v_size;
```

Comment:
maybe instead:

```c
info_len = mempcpy(mempcpy(stpcpy(info, label) + 1, 
                                  context_u, context_u_size),
                                  context_v, context_vsize))) - info;
```

you get the idea

Reply:
yep thanks, and i added an assert to verify the amount copied was correct